File: ui/desktop/src/acp/adapter/toolNotifications.ts
import type { ToolCallUpdate } from '@agentclientprotocol/sdk';
import type { NotificationEvent } from '../../types/message';
import type { AcpChatStateChange } from './shared';
import { isRecord } from './shared';

type ToolNotification =
  | {
      type: 'message';
      params: LoggingMessageNotificationParams;
    }
  | {
      type: 'progress';
      params: ProgressNotificationParams;
    }
  | {
      type: 'platform_event';
      params: PlatformEventParams;
    };

type LoggingMessageNotificationParams = {
  level: string;
  logger?: string;
  data: unknown;
};

type ProgressNotificationParams = {
  progressToken: string | number;
  progress: number;
  total?: number;
  message?: string;
};

type PlatformEventParams = Record<string, unknown>;

export function toolNotificationChange(
  update: ToolCallUpdate
): Extract<AcpChatStateChange, { type: 'notification' }> | undefined {
  const notification = toolNotificationEvent(update);
  if (!notification) {
    return undefined;
  }

  return {
    type: 'notification',
    notification,
  };
}

export function toolNotificationEvent(update: ToolCallUpdate): NotificationEvent | undefined {
  const toolNotification = parseToolNotification(update._meta);
  if (!toolNotification) {
    return undefined;
  }

  return toNotificationEvent(update.toolCallId, toolNotification);
}

function parseToolNotification(meta: unknown): ToolNotification | undefined {
  if (!isRecord(meta)) {
    return undefined;
  }

  const toolNotification = meta.toolNotification;
  if (!isRecord(toolNotification)) {
    return undefined;
  }

  if (toolNotification.type === 'message') {
    const params = parseLoggingMessageParams(toolNotification.params);
    return params ? { type: 'message', params } : undefined;
  }

  if (toolNotification.type === 'progress') {
    const params = parseProgressParams(toolNotification.params);
    return params ? { type: 'progress', params } : undefined;
  }

  if (toolNotification.type === 'platform_event') {
    const params = parsePlatformEventParams(toolNotification.params);
    return params ? { type: 'platform_event', params } : undefined;
  }

  return undefined;
}

function parseLoggingMessageParams(value: unknown): LoggingMessageNotificationParams | undefined {
  if (!isRecord(value) || typeof value.level !== 'string' || !('data' in value)) {
    return undefined;
  }

  return {
    level: value.level,
    ...(typeof value.logger === 'string' ? { logger: value.logger } : {}),
    data: value.data,
  };
}

function parseProgressParams(value: unknown): ProgressNotificationParams | undefined {
  if (
    !isRecord(value) ||
    (typeof value.progressToken !== 'string' && typeof value.progressToken !== 'number') ||
    typeof value.progress !== 'number'
  ) {
    return undefined;
  }

  return {
    progressToken: value.progressToken,
    progress: value.progress,
    ...(typeof value.total === 'number' ? { total: value.total } : {}),
    ...(typeof value.message === 'string' ? { message: value.message } : {}),
  };
}

function parsePlatformEventParams(value: unknown): PlatformEventParams | undefined {
  return isRecord(value) ? value : undefined;
}

function toNotificationEvent(
  toolCallId: string,
  toolNotification: ToolNotification
): NotificationEvent {
  return {
    type: 'Notification',
    request_id: toolCallId,
    message: {
      method: notificationMethod(toolNotification),
      params: toolNotification.params,
    },
  };
}

function notificationMethod(toolNotification: ToolNotification): string {
  switch (toolNotification.type) {
    case 'message':
      return 'notifications/message';
    case 'progress':
      return 'notifications/progress';
    case 'platform_event':
      return 'platform_event';
  }
}

File: crates/goose/src/acp/server/tool_notifications.rs
use agent_client_protocol::schema::{
    Meta, SessionUpdate, ToolCallId, ToolCallStatus, ToolCallUpdate, ToolCallUpdateFields,
};
use rmcp::model::{LoggingMessageNotificationParam, ProgressNotificationParam, ServerNotification};
use serde::Serialize;

#[derive(Debug, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
enum ToolNotification {
    Message {
        params: LoggingMessageNotificationParam,
    },
    Progress {
        params: ProgressNotificationParam,
    },
    PlatformEvent {
        params: serde_json::Value,
    },
}

pub(super) fn tool_notification_update(
    tool_call_id: impl Into<ToolCallId>,
    notification: ServerNotification,
) -> Option<SessionUpdate> {
    let tool_notification = match notification {
        ServerNotification::LoggingMessageNotification(notification) => ToolNotification::Message {
            params: notification.params,
        },
        ServerNotification::ProgressNotification(notification) => ToolNotification::Progress {
            params: notification.params,
        },
        ServerNotification::CustomNotification(notification)
            if notification.method == "platform_event" =>
        {
            ToolNotification::PlatformEvent {
                params: notification.params.unwrap_or(serde_json::Value::Null),
            }
        }
        _ => return None,
    };

    let mut meta = Meta::new();
    meta.insert(
        "toolNotification".to_string(),
        serde_json::to_value(tool_notification).ok()?,
    );

    Some(SessionUpdate::ToolCallUpdate(
        ToolCallUpdate::new(
            tool_call_id,
            ToolCallUpdateFields::new().status(ToolCallStatus::InProgress),
        )
        .meta(meta),
    ))
}

#[cfg(test)]
mod tests {
    use super::tool_notification_update;
    use rmcp::model::{
        CancelledNotificationParam, CustomNotification, LoggingLevel,
        LoggingMessageNotificationParam, Notification, NumberOrString, ProgressNotificationParam,
        ProgressToken, ServerNotification,
    };
    use serde_json::json;
    use std::sync::Arc;

    #[test]
    fn maps_logging_message_notification_to_tool_update_meta() {
        let notification = ServerNotification::LoggingMessageNotification(Notification::new(
            LoggingMessageNotificationParam::new(
                LoggingLevel::Info,
                json!({
                    "type": "subagent_tool_request",
                    "subagent_id": "session_1",
                    "tool_call": {
                        "name": "developer__shell"
                    }
                }),
            )
            .with_logger("subagent:session_1"),
        ));

        let update = tool_notification_update("tool_1", notification).expect("expected update");
        let value = serde_json::to_value(update).expect("update should serialize");

        assert_eq!(value["sessionUpdate"], "tool_call_update");
        assert_eq!(value["toolCallId"], "tool_1");
        assert_eq!(value["status"], "in_progress");
        assert_eq!(value["_meta"]["toolNotification"]["type"], "message");
        assert_eq!(
            value["_meta"]["toolNotification"]["params"]["level"],
            "info"
        );
        assert_eq!(
            value["_meta"]["toolNotification"]["params"]["logger"],
            "subagent:session_1"
        );
        assert_eq!(
            value["_meta"]["toolNotification"]["params"]["data"]["tool_call"]["name"],
            "developer__shell"
        );
    }

    #[test]
    fn maps_progress_notification_to_tool_update_meta() {
        let notification = ServerNotification::ProgressNotification(Notification::new(
            ProgressNotificationParam::new(
                ProgressToken(NumberOrString::String(Arc::from("scan-repo"))),
                3.0,
            )
            .with_total(10.0)
            .with_message("Scanned 3 of 10 directories"),
        ));

        let update = tool_notification_update("tool_1", notification).expect("expected update");
        let value = serde_json::to_value(update).expect("update should serialize");

        assert_eq!(value["sessionUpdate"], "tool_call_update");
        assert_eq!(value["toolCallId"], "tool_1");
        assert_eq!(value["status"], "in_progress");
        assert_eq!(value["_meta"]["toolNotification"]["type"], "progress");
        assert_eq!(
            value["_meta"]["toolNotification"]["params"]["progressToken"],
            "scan-repo"
        );
        assert_eq!(
            value["_meta"]["toolNotification"]["params"]["progress"],
            3.0
        );
        assert_eq!(value["_meta"]["toolNotification"]["params"]["total"], 10.0);
        assert_eq!(
            value["_meta"]["toolNotification"]["params"]["message"],
            "Scanned 3 of 10 directories"
        );
    }

    #[test]
    fn ignores_non_tool_live_notification_variants() {
        let notification = ServerNotification::CancelledNotification(Notification::new(
            CancelledNotificationParam {
                request_id: NumberOrString::String(Arc::from("request_1")),
                reason: None,
            },
        ));

        assert!(tool_notification_update("tool_1", notification).is_none());
    }

    #[test]
    fn maps_platform_event_custom_notification_to_tool_update_meta() {
        let notification = ServerNotification::CustomNotification(CustomNotification::new(
            "platform_event",
            Some(json!({
                "extension": "apps",
                "event_type": "app_created",
                "app_name": "platform-event-repro"
            })),
        ));

        let update = tool_notification_update("tool_1", notification).expect("expected update");
        let value = serde_json::to_value(update).expect("update should serialize");

        assert_eq!(value["sessionUpdate"], "tool_call_update");
        assert_eq!(value["toolCallId"], "tool_1");
        assert_eq!(value["status"], "in_progress");
        assert_eq!(value["_meta"]["toolNotification"]["type"], "platform_event");
        assert_eq!(
            value["_meta"]["toolNotification"]["params"]["extension"],
            "apps"
        );
        assert_eq!(
            value["_meta"]["toolNotification"]["params"]["event_type"],
            "app_created"
        );
        assert_eq!(
            value["_meta"]["toolNotification"]["params"]["app_name"],
            "platform-event-repro"
        );
    }

    #[test]
    fn ignores_non_platform_event_custom_notifications() {
        let notification = ServerNotification::CustomNotification(CustomNotification::new(
            "notifications/custom",
            Some(json!({ "extension": "apps" })),
        ));

        assert!(tool_notification_update("tool_1", notification).is_none());
    }
}
